crates/ruff/src/jupyter/notebook.rs

```diff
@@ -37,7 +29,53 @@ pub fn is_jupyter_notebook(path: &Path) -> bool {
         && cfg!(feature = "jupyter_notebook")
 }
 
-impl JupyterNotebook {
+/// Return `true` if the given [`Cell`] is a valid code cell.
+///
+/// A valid code cell is a cell where the type is [`CellType::Code`] and the
+/// source is not a magic, shell or help command.
+fn is_valid_code_cell(cell: &Cell) -> bool {
+    if cell.cell_type != CellType::Code {
+        return false;
+    }
+    let lines = match &cell.source {
+        SourceValue::String(string) => string.lines().collect::<Vec<_>>(),
+        SourceValue::StringArray(string_array) => string_array
+            .iter()
+            .map(std::string::String::as_str)
+            .collect(),
+    };
+    // Ignore a cell if it contains a magic command. There could be valid
+    // Python code as well, but we'll ignore that for now.
+    // TODO(dhruvmanila): https://github.com/psf/black/blob/main/src/black/handle_ipynb_magics.py
+    !lines.iter().any(|line| {
+        MAGIC_PREFIX
+            .iter()
+            .any(|prefix| line.trim_start().starts_with(prefix))
+    })
+}
+
+#[derive(Debug)]
+pub struct Notebook {
+    /// Python source code of the notebook.
+    ///
+    /// This is the concatenation of all valid code cells in the notebook
+    /// separated by a newline and a trailing newline. The trailing newline
+    /// is added to make sure that each cell ends with a newline which will
+    /// be removed when updating the cell content.
+    content: String,
+    /// The index of the notebook. This is used to map between the concatenated
+    /// source code and the original notebook.
+    index: JupyterIndex,
+    /// The raw notebook.
+    raw: JupyterNotebook,
+    /// The offsets of each cell in the concatenated source code. This includes
+    /// the first and last character offsets as well.
+    cell_offsets: Vec<TextSize>,
+    /// The cell numbers of all valid code cells in the notebook.
+    valid_code_cells: Vec<usize>,
```

Comment:
Nit: Would a `u32` suffice?

How is the cell number defined?

Reply:
Yes, it should 👍 

> How is the cell number defined?

Oh, actually it's an index value in a `Vec<Cell>` which are available on `JupyterNotebook`

Reply:
Would it be possible to use our `IndexVec` and `newtype_index` macro to make this relationship explicit? 

https://github.com/charliermarsh/ruff/blob/d1c7a07d2ae56f4ff56b81c7f2a398e2262c42b9/crates/ruff_python_semantic/src/scope.rs#L140-L163

Reply:
I'm not sure if that'll work, correct me if I'm wrong. So, `IndexVec` will own it's elements. This means we'll have to copy all of the valid code `Cell` objects into this vector. At the end, we'll have to copy it back to the `RawNotebook` to serialize it into JSON. Move won't work here as only valid cells will be part of the vector thus leaving a partial vector.